core/src/main/java/org/infinispan/util/ExponentialBackOff.java

```diff
@@ -0,0 +1,54 @@
+package org.infinispan.util;
+
+import org.infinispan.commons.util.Experimental;
+
+/**
+ * Interface to implement an exponential back-off algorithm that retries the request based on the result of the remote
+ * operation.
+ * <p>
+ * Currently only used for IRAC, this interface contains 2 methods: {@link #backoffSleep()} which should be invoked if
+ * the request needs to be retried and {@link #reset()}, invoked when a request is "successful", which resets the
+ * state.
+ * <p>
+ * The interface may me changed in the future to include async methods.
+ *
+ * @author Pedro Ruivo
+ * @since 12.0
+ */
+@Experimental
+public interface ExponentialBackOff {
+
+   /**
+    * Disabled exponential back-off algorithm. It does nothing.
+    */
+   ExponentialBackOff NO_OP = new ExponentialBackOff() {
+      @Override
+      public void backoffSleep() {
+         //no-op
+      }
+
+      @Override
+      public void reset() {
+         //no-op
+      }
+   };
+
+   /**
+    * It blocks the thread for a certain amount of time before retries the request.
+    * <p>
+    * The method is blocking and should be invoked when a request needs to be retried. It blocks the thread for a
```

Comment:
I am not a fan of this. Can we not instead use the `TIMEOUT_SCHEDULE_EXECUTOR` executor to do this? Or is it required to stay in the invoking thread for preserveOrder ?

Reply:
this is on the sending. IRAC has a dedicated thread looping through the changed keys. Sleep if fine here since the thread isn't used for anything else.

Reply:
Okay, tbh I was even going to discuss possibly removing the IRAC thread at some point and just submit as needed and/or use the scheduled executor so we don't need an additional thread. However it isn't worth it to do in this PR it seems.

Reply:
My idea is to use the `BlockingManager` directly. Use one `BlockingThread` to send the requests (since they may involve persistence) could save some thread spawning. Also, I need to avoid spawning a thread if a batch is in progress. Something like this:

1. Spawn single `BlockThread` and block more thread spawning 
2. Send all the requests and return the thread to `BlockingManager`
3. When all requests finish, spawn a `BlockThread` to handle the replies
a. if not more keys to send, allow thread spawning and return the thread to `BlockingManager`
b. else, send the next batch (goto 2)